src/wire/dhcp.rs

```diff
@@ -0,0 +1,632 @@
+    pub const OPT_TCP_KEEPALIVE_INTERVAL: u8 = 38;
+    pub const OPT_TCP_KEEPALIVE_GARBAGE: u8 = 39;
+
+    // Application and Service Parameters
+    pub const OPT_NIS_DOMAIN: u8 = 40;
+    pub const OPT_NIS_SERVERS: u8 = 41;
+    pub const OPT_NTP_SERVERS: u8 = 42;
+    pub const OPT_VENDOR_SPECIFIC_INFO: u8 = 44;
+    pub const OPT_NETBIOS_SERVER: u8 = 45;
+    pub const OPT_NETBIOS_NODE_TYPE: u8 = 46;
+    pub const OPT_NETBIOS_SCOPE: u8 = 47;
+    pub const OPT_X_WINDOW_FONT_SERVER: u8 = 48;
+    pub const OPT_X_WINDOW_DISPLAY_MANAGER: u8 = 49;
+    pub const OPT_NIS_PLUS_DOMAIN: u8 = 64;
+    pub const OPT_NIS_PLUS_SERVERS: u8 = 65;
+    pub const OPT_MOBILE_IP_HOME_AGENT: u8 = 68;
+    pub const OPT_SMTP_SERVER: u8 = 69;
+    pub const OPT_POP3_SERVER: u8 = 70;
+    pub const OPT_NNTP_SERVER: u8 = 71;
+    pub const OPT_WWW_SERVER: u8 = 72;
+    pub const OPT_FINGER_SERVER: u8 = 73;
+    pub const OPT_IRC_SERVER: u8 = 74;
+    pub const OPT_STREETTALK_SERVER: u8 = 75;
+    pub const OPT_STDA_SERVER: u8 = 76;
+
+    // DHCP Extensions
+    pub const OPT_REQUESTED_IP: u8 = 50;
+    pub const OPT_IP_LEASE_TIME: u8 = 51;
+    pub const OPT_OPTION_OVERLOAD: u8 = 52;
+    pub const OPT_TFTP_SERVER_NAME: u8 = 66;
+    pub const OPT_BOOTFILE_NAME: u8 = 67;
+    pub const OPT_DHCP_MESSAGE_TYPE: u8 = 53;
+    pub const OPT_SERVER_IDENTIFIER: u8 = 54;
+    pub const OPT_PARAMETER_REQUEST_LIST: u8 = 55;
+    pub const OPT_MESSAGE: u8 = 56;
+    pub const OPT_MAX_DHCP_MESSAGE_SIZE: u8 = 57;
+    pub const OPT_RENEWAL_TIME_VALUE: u8 = 58;
+    pub const OPT_REBINDING_TIME_VALUE: u8 = 59;
+    pub const OPT_VENDOR_CLASS_ID: u8 = 60;
+    pub const OPT_CLIENT_ID: u8 = 61;
+}
+
+impl<T: AsRef<[u8]>> Packet<T> {
+    /// Imbue a raw octet buffer with DHCP packet structure.
+    pub fn new(buffer: T) -> Packet<T> {
+        Packet { buffer }
+    }
+
+    /// Shorthand for a combination of [new] and [check_len].
+    ///
+    /// [new]: #method.new
+    /// [check_len]: #method.check_len
+    pub fn new_checked(buffer: T) -> Result<Packet<T>> {
+        let packet = Self::new(buffer);
+        packet.check_len()?;
+        Ok(packet)
+    }
+
+    /// Ensure that no accessor method will panic if called.
+    /// Returns `Err(Error::Truncated)` if the buffer is too short.
+    ///
+    /// [set_header_len]: #method.set_header_len
+    pub fn check_len(&self) -> Result<()> {
+        let len = self.buffer.as_ref().len();
+        if len < field::MAGIC_NUMBER.end {
+            Err(Error::Truncated)
+        } else {
+            Ok(())
+        }
+    }
+
+    /// Consume the packet, returning the underlying buffer.
+    pub fn into_inner(self) -> T {
+        self.buffer
+    }
+
+    /// Returns wether this packet is a request (1) or a reply (2).
+    pub fn opcode(&self) -> u8 {
+        let data = self.buffer.as_ref();
+        data[field::OP]
+    }
+
+    pub fn hardware_address_type(&self) -> HardwareAddressType {
```

Comment:
The names of this and the following method as well as their values should be shared with ARP.